crates/ruff_python_formatter/src/lib.rs

```diff
@@ -57,70 +55,98 @@ pub fn fmt(contents: &str) -> Result<Formatted<ASTFormatContext>> {
-        insta::assert_display_snapshot!(snapshot, formatted.print()?.as_code());
-        Ok(())
-    }
 
-    #[test_case(Path::new("simple_cases/collections.py"); "collections")]
-    #[test_case(Path::new("simple_cases/bracketmatch.py"); "bracketmatch")]
-    fn passing_modulo_string_normalization(path: &Path) -> Result<()> {
-        fn adjust_quotes(contents: &str) -> String {
-            // Replace all single quotes with double quotes.
-            contents.replace('\'', "\"")
+        let expected_path = input_path.with_extension("py.expect");
+        let expected_output = fs::read_to_string(&expected_path)
+            .unwrap_or_else(|_| panic!("Expected Black output file '{expected_path:?}' to exist"));
+
+        let printed = formatted.print()?;
+        let formatted_code = printed.as_code();
+
+        if formatted_code == expected_output {
+            // Black and Ruff formatting matches. Delete any existing snapshot files because the Black output
+            // already perfectly captures the expected output.
+            // The following code mimics insta's logic generating the snapshot name for a test.
+            let workspace_path = std::env::var("CARGO_MANIFEST_DIR").unwrap();
+            let snapshot_name = insta::_function_name!()
+                .strip_prefix(&format!("{}::", module_path!()))
+                .unwrap();
+            let module_path = module_path!().replace("::", "__");
+
+            let snapshot_path = Path::new(&workspace_path)
+                .join("src/snapshots")
+                .join(&format!(
+                    "{module_path}__{}.snap",
+                    snapshot_name.replace(&['/', '\\'][..], "__")
+                ));
+
+            if snapshot_path.exists() && snapshot_path.is_file() {
+                // SAFETY: This is a convenience feature. That's why we don't want to abort
+                // when deleting a no longer needed snapshot fails.
+                fs::remove_file(&snapshot_path).ok();
+            }
+
+            let new_snapshot_path = snapshot_path.with_extension("snap.new");
+            if new_snapshot_path.exists() && new_snapshot_path.is_file() {
+                // SAFETY: This is a convenience feature. That's why we don't want to abort
+                // when deleting a no longer needed snapshot fails.
+                fs::remove_file(&new_snapshot_path).ok();
+            }
+        } else {
+            // Black and Ruff have different formatting. Write out a snapshot that covers the differences
+            // today.
+            let mut snapshot = String::new();
+            write!(snapshot, "{}", Header::new("Input"))?;
+            write!(snapshot, "{}", CodeFrame::new("py", &content))?;
+
+            write!(snapshot, "{}", Header::new("Black Differences"))?;
+
+            let diff = TextDiff::from_lines(expected_output.as_str(), formatted_code)
+                .unified_diff()
+                .header("Black", "Ruff")
+                .to_string();
```

Comment:
That's correct. I found it useful when working on rome formatter to have all information in a single file, so that you don't need to search for the expected output file.